crates/uv/src/commands/project/mod.rs

```diff
@@ -81,42 +82,94 @@ pub(crate) fn find_environment(
 
-            writeln!(
-                printer.stderr(),
-                "Using Python {} interpreter at: {}",
-                interpreter.python_version(),
-                interpreter.sys_executable().user_display().cyan()
-            )?;
+    let requires_python = workspace
+        .root_member()
+        .and_then(|root| root.project().requires_python.as_ref());
+
+    // Discover or create the virtual environment.
+    match PythonEnvironment::from_root(venv, cache) {
+        Ok(venv) => {
+            // `--python` has highest precedence, after that we check `requires_python` from
+            // `pyproject.toml`. If `--python` and `requires_python` are mutually incompatible,
+            // we'll fail at the build or at last the install step when we aren't able to install
+            // the editable wheel for the current project into the venv.
+            // TODO(konsti): Do we want to support a workspace python version requirement?
+
+            let venv_python_satisfactory = if let Some(python) = python {
+                ToolchainRequest::parse(python).satisfied(venv.interpreter())
+            } else if let Some(requires_python) = requires_python {
+                requires_python.contains(venv.interpreter().python_version())
+            } else {
+                true
+            };
+
+            if venv_python_satisfactory {
+                return Ok(venv);
+            }
 
-            let venv = workspace.venv();
             writeln!(
                 printer.stderr(),
-                "Creating virtualenv at: {}",
-                venv.user_display().cyan()
+                "Removing virtual environment at: {}",
+                venv.root().user_display().cyan()
             )?;
+            fs_err::remove_dir_all(venv.root())
+                .context("Failed to remove existing virtual environment")?;
+        }
+        Err(uv_toolchain::Error::NotFound(_)) => {}
+        Err(e) => return Err(e.into()),
+    }
 
-            Ok(uv_virtualenv::create_venv(
-                &venv,
-                interpreter,
-                uv_virtualenv::Prompt::None,
-                false,
-                false,
-            )?)
+    // TODO(konsti): Extend `VersionRequest` to support `VersionSpecifiers`.
+    let requires_python_str = requires_python.map(ToString::to_string);
+    let interpreter = Toolchain::find(
+        python.or(requires_python_str.as_deref()),
+        // Otherwise we'll try to use the venv we just deleted.
+        SystemPython::Required,
+        preview,
+        cache,
+    )?
+    .into_interpreter();
+
+    if let Some(requires_python) = requires_python {
+        if !requires_python.contains(interpreter.python_version()) {
+            warn_user!(
+                "The Python {} you requested with {} is incompatible with the requirement of the \
+                project of {}",
+                interpreter.python_version(),
+                python.unwrap_or("(default)"),
```

Comment:
Can we `if`-`else` this warning? `you requested with (default) is incompatible...` is confusing and it's trivial for us to use a better message here.